### PR TITLE
tarindex: Add special symlink name handling

### PR DESCRIPTION
**Pull request by miz060**
###### Merge Checklist  

- [x] Followed patch format from upstream recommendation: https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md#patch-format
  - [x] Included a single commit in a given PR - at least unless there are related commits and each makes sense as a change on its own.
- [x] Aware about the PR to be merged using "create a merge commit" rather than "squash and merge" (or similar)
- [x] genPolicy only: Ensured the tool still builds on Windows
- [x] genPolicy only: Updated sample YAMLs' policy annotations, if applicable
- [x] The `upstream-missing` label (or `upstream-not-needed`) has been set on the PR. 

###### Summary 

Enable tarindex to handle symlink name with special characters or length longer than the 100 limit by duplicating it before the string table.

###### Test Methodology

- Local testing with tarindex execuatable.
- Manual AKS cluster testing with "registry.k8s.io/e2e-test-images/agnhost:2.40" image. This update creates the previously skipped `/etc/ssl/certs/ca-cert-NetLock_Arany_=Class_Gold=_Főtanúsítvány.pem symlink`.
- RPM build: [510540](https://dev.azure.com/mariner-org/mariner/_build/results?buildId=510540&view=results)
- Image build: [510563](https://dev.azure.com/mariner-org/mariner/_build/results?buildId=510563&view=results)
- Conformance test: [510729](https://dev.azure.com/mariner-org/mariner/_build/results?buildId=510729&view=results)
- Performance test: [510730](https://dev.azure.com/mariner-org/mariner/_build/results?buildId=510730&view=results)

### CODE REVIEW

**Comment by miz060** [comment]
> The change needs to target our fork because the part of the code we fix is not upstreamed yet?

That's correct.

File: src/tools/genpolicy/src/volume.rs [hunk 1]
```diff
@@ -37,8 +37,7 @@ pub struct Volume {
     pub secret: Option<SecretVolumeSource>,
 
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub downwardAPI: Option<DownwardAPIVolumeSource>
-    // TODO: additional fields.
+    pub downwardAPI: Option<DownwardAPIVolumeSource>, // TODO: additional fields.
```
Comment (danmihai1) [inline src/tools/genpolicy/src/volume.rs:40]: Is this change required?

**Review by danmihai1 — CHANGES_REQUESTED** [review]

File: src/tools/genpolicy/src/volume.rs (hunk 1 above)
Comment (danmihai1) [reply src/tools/genpolicy/src/volume.rs:40]: Please use a separate PR for any such unrelated changes.

Comment (miz060) [reply src/tools/genpolicy/src/volume.rs]: Got it. Changes reverted.

**Review by danmihai1 — APPROVED** [review]

**Review by ms-mahuber — APPROVED** [review]

**Review by sprt — CHANGES_REQUESTED** [review]
Could you squash your commits into one?

File: src/tardev-snapshotter/tarindex/src/lib.rs [hunk 2]
```diff
@@ -447,14 +455,20 @@ pub fn append_index(data: &mut (impl io::Read + io::Write + io::Seek)) -> io::Re
             group: e.group.into(),
             lmtime: (e.mtime as u32).into(),
             size: e.size.into(),
-            offset: e.offset.into(),
+            offset: inode_offset,
```
Comment (sprt) [inline src/tardev-snapshotter/tarindex/src/lib.rs:458]: nit: call `.into()` here instead of up there - more readable IMO

**Comment by miz060** [comment]
> Could you squash your commits into one?

Yes, squashed now.

File: src/tardev-snapshotter/tarindex/src/lib.rs (hunk 2 above)
Comment (miz060) [reply src/tardev-snapshotter/tarindex/src/lib.rs]: good point. updated.

File: src/tardev-snapshotter/tarindex/src/lib.rs [hunk 3]
```diff
@@ -136,18 +137,13 @@ fn read_all_entries(
...
-                                "Skipping symlink with long link name ({}, {} bytes, {}, {} bytes): {}",
-                                String::from_utf8_lossy(&name), name.len(),
-                                String::from_utf8_lossy(&hname), hname.len(),
-                                String::from_utf8_lossy(&f.path_bytes())
-                            );
-                            continue;
+                            special_link.push(String::from_utf8_lossy(&name).to_string());
+                            entry_offset = 0;
+                        }
+                        else {
```
Comment (wedsonaf) [inline src/tardev-snapshotter/tarindex/src/lib.rs:143]: Did you run cargo fmt? I think it would merge the two lines above into a single.

File: src/tardev-snapshotter/tarindex/src/lib.rs [hunk 4]
```diff
@@ -427,13 +424,24 @@ pub fn append_index(data: &mut (impl io::Read + io::Write + io::Seek)) -> io::Re
...
         }
 
+        // Check for special symlink names
+        let inode_offset = if (e.mode & S_IFMT) == S_IFLNK &&  e.offset == 0 {
+            symlink_offset
+        } else {
+            e.offset
+        };
+        if (e.mode & S_IFMT) == S_IFLNK &&  e.offset == 0 {
+            symlink_offset += e.size;
```
Comment (wedsonaf) [inline src/tardev-snapshotter/tarindex/src/lib.rs:442]: To avoid having to repeat this check, I'd do this on line 437 above.

Something like:
```
    let inode_offset = if (e.mode & S_IFMT) == S_IFLNK && e.offset == 0 {
            let v = symlink_offset;
            symlink_offset += e.size;
            v
        } else {
            e.offset
        }.into();
```

In fact, it's preferable to leave the short end at the top, so I would change to:
```
    let inode_offset = if (e.mode & S_IFMT) != S_IFLNK || e.offset != 0 {
            e.offset
        } else {
            let v = symlink_offset;
            symlink_offset += e.size;
            v
        }.into();
```

File: src/tardev-snapshotter/tarindex/src/lib.rs (hunk 3 above)
Comment (miz060) [reply src/tardev-snapshotter/tarindex/src/lib.rs]: Good catch! I forgot to run cargo fmt. Just pushed its format changes.

File: src/tardev-snapshotter/tarindex/src/lib.rs (hunk 4 above)
Comment (miz060) [reply src/tardev-snapshotter/tarindex/src/lib.rs:439]: Thanks for the suggestion! Updated with style of the second block.

**Review by sprt — APPROVED** [review]

File: src/tardev-snapshotter/tarindex/src/lib.rs [hunk 5]
```diff
@@ -306,7 +301,7 @@ pub fn append_index(data: &mut (impl io::Read + io::Write + io::Seek)) -> io::Re
         mode: S_IFDIR | 0o555,
         ..Entry::default()
     }));
-
+    let mut special_link: Vec<Vec<u8>> = Vec::new();
```
Comment (wedsonaf) [inline src/tardev-snapshotter/tarindex/src/lib.rs:304]: nit: I believe the type can be inferred here, so it can probably be omitted.

File: src/tardev-snapshotter/tarindex/src/lib.rs [hunk 6]
```diff
@@ -136,18 +137,12 @@ fn read_all_entries(
...
                         if *hname != *name {
-                            // TODO: Handle this case by duplicating the full name.
-                            eprintln!(
-                                "Skipping symlink with long link name ({}, {} bytes, {}, {} bytes): {}",
-                                String::from_utf8_lossy(&name), name.len(),
-                                String::from_utf8_lossy(&hname), hname.len(),
-                                String::from_utf8_lossy(&f.path_bytes())
-                            );
-                            continue;
+                            special_link.push((&name).to_vec());
```
Comment (wedsonaf) [inline src/tardev-snapshotter/tarindex/src/lib.rs:140]: Can't it just be `name.to_vec()`? That is, omitting parentheses and ampersand?

**Review by wedsonaf — APPROVED** [review]
LGTM. Left a couple of nits.

File: src/tardev-snapshotter/tarindex/src/lib.rs (hunk 5 above)
Comment (miz060) [reply src/tardev-snapshotter/tarindex/src/lib.rs]: Yes. type omitted.

File: src/tardev-snapshotter/tarindex/src/lib.rs [hunk 7]
```diff
@@ -51,6 +51,7 @@ fn read_all_entries(
     root: &mut Rc<RefCell<Entry>>,
     mut cb: impl FnMut(&mut Rc<RefCell<Entry>>, &[u8], &Entry),
     mut hardlink: impl FnMut(&mut Rc<RefCell<Entry>>, &[u8], &[u8]),
+    special_link: &mut Vec<Vec<u8>>,
```
[the remark this answers is not in the document]
Comment (miz060) [reply src/tardev-snapshotter/tarindex/src/lib.rs]: Good point! Moved `special_link` as suggested.

File: src/tardev-snapshotter/tarindex/src/lib.rs (hunk 6 above)
Comment (miz060) [reply src/tardev-snapshotter/tarindex/src/lib.rs]: Good catch! Yes. Code simplified.

File: src/agent/rustjail/src/container.rs [hunk 8]
```diff
@@ -206,7 +206,7 @@ lazy_static! {
                 gid: Some(0xffffffff),
             },
         ];
-
```
[the remark this answers is not in the document]
Comment (miz060) [reply src/agent/rustjail/src/container.rs]: It's an unintentional auto-format change. Reverted.
